feat: add new 'ibm-no-ref-in-example' rule

padamstx opened the pull request:
## PR summary
This commit introduces the new 'ibm-no-ref-in-example' rule which will check to make sure that "$ref" is not used in "example" fields.  Note that $ref is valid in an "examples" field, but not in an "example" field.

## PR Checklist

### General checklist
Please make sure that your PR fulfills the following requirements:  
- [x] The commit message follows the [Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] Dependencies have been updated as needed
- [ ] .secrets.baseline updated as needed?

#### Checklist for adding a new validation rule:
- [x] Added new validation rule definition (packages/ruleset/src/rules/*.js, index.js)
- [x] If necessary, added new validation rule implementation (packages/ruleset/src/functions/*.js, updated index.js)
- [x] Added new rule to default configuration (packages/ruleset/src/ibm-oas.js)
- [x] Added tests for new rule (packages/ruleset/test/*.test.js)
- [x] Added docs for new rule (docs/ibm-cloud-rules.md)

HenryGessau commented on packages/ruleset/test/no-ref-in-example.test.js:
```diff
@@ -0,0 +1,202 @@
+            schema: {
+              type: 'string',
+            },
+            example: {
+              $ref: '#/components/schemas/Error',
```
Should we also have a few test cases with more deeply nested ref's?
E..g.:
```yaml
example:
  library:
    name: my-library
    address:
      $ref: '#/components/schemas/Address/example'
    books:
    - $ref: '#/components/schemas/Book/example'
```

dpopp07 replied:
This is tricky because I don't think that's _technically_ invalid. It's just probably not what the user wants.

For example, the above _could_ be a representation of this valid JSON object:
```json
"library": {
  "name": "my-library",
  "address": {
    "$ref": "#/components/schemas/Address/example"
  },
  "books": [{ "$ref": "#/components/schemas/Book/example" }]
}
```

where `"$ref"` is just an arbitrary key with no meaning. I know that seems rare and maybe we could warn to remind the user its _likely_ wrong but it's hard to tell the difference between unintentional and intentional use.

padamstx replied:
@HenryGessau I can understand the reason for your comment, and this scenario certainly has some subtleties associated with it.  But, I agree with Dustin here in that an argument could be made that the example object you have posted is probably valid.  I think the rule should be focused on looking for a top-level "$ref" property within the "example" object, and avoid checking for nested uses of "$ref" since those might be valid.

meem replied:
FWIW, our hacky check found several places in the VPC spec where spec authors have done exactly what Henry's asking about: https://github.ibm.com/riaas/api-spec/pull/6260/files

padamstx replied:
Ok, I'm convinced :joy:
I'll expand the scope of the rule so that it flags a `$ref` property within the `example` field at any level of nesting.

padamstx replied:
fixed

dpopp07 commented on docs/ibm-cloud-rules.md:
```diff
@@ -63,7 +63,9 @@ which is delivered in the `@ibm-cloud/openapi-ruleset` NPM package.
   * [ibm-no-optional-properties-in-required-body](#ibm-no-optional-properties-in-required-body)
+  * [ibm-no-ref-in-example](#ibm-no-ref-in-example)
   * [ibm-no-space-in-example-name](#ibm-no-space-in-example-name)
+  * [ibm-no-superfluous-allof](#ibm-no-superfluous-allof)
```
Nice catch!

dpopp07 commented on packages/ruleset/src/rules/no-ref-in-example.js:
```diff
@@ -0,0 +1,19 @@
+ */
+
+const { oas3 } = require('@stoplight/spectral-formats');
+// const { falsy } = require('@stoplight/spectral-functions');
```
Can be removed

padamstx replied:
fixed

dpopp07 reviewed (approved):
Looks good! 👍

padamstx commented:
@dpopp07 @HenryGessau @meem
I've updated the rule function so that it performs the search for a `$ref` field recursively, and I've updated the tests accordingly.

dpopp07 reviewed (approved):
Updated code looks good! 👍

ibm-devx-sdk commented:
:tada: This PR is included in version 1.18.0 :tada:

The release is available on [npm package (@latest dist-tag)](https://www.npmjs.com/package/@ibm-cloud/openapi-ruleset/v/1.18.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

ibm-devx-sdk commented:
:tada: This PR is included in version 1.4.0 :tada:

The release is available on [npm package (@latest dist-tag)](https://www.npmjs.com/package/@ibm-cloud/openapi-ruleset-utilities/v/1.4.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
